### ISSUE TITLE
serve s3: let rclone act as an S3 compatible server serving any backend

### ISSUE THREAD

**Issue by jkaberg** [rclone/rclone#3382]
### What is your current rclone version (output from `rclone version`)?
```
rclone v1.47.0
- os/arch: linux/amd64
- go version: go1.12.4
```

#### What problem are you are trying to solve?
Serve S3 protocol for backends that don't support it

#### How do you think rclone should be changed to solve that?
Let rclone via the serve parameter act as an "s3 proxy". Currently rclone supports many backends which lacks this, while many other tools commonly implements S3 as a backend. Something like https://github.com/minio/minio coupled with rclone backend's would be very cool. Yes I could do it with fuse+minio, however that would add another tool and complexity.

**Comment by ncw** [comment]
I really like this idea.  The obvious thing would be to take the internals of minio and use them as a library, however last time I looked at its code it wasn't designed for that.

It might be worth another look...

**Comment by jkaberg** [comment]
Thanks for considering @ncw, afaik there aren't many golang compatible implementations to the [official API](https://docs.aws.amazon.com/AmazonS3/latest/API/Welcome.html) that I know

- https://github.com/minio/minio (100% compatible?)
- https://github.com/johannesboyne/gofakes3 (missing features)

There seems to be an minio issue tracking an more library/package approach over [here](https://github.com/minio/minio/issues/4812)

**Comment by ncw** [comment]
Gofakes looks quite promising, eg the [backend definiton](https://godoc.org/github.com/johannesboyne/gofakes3#Backend).

It would be possible with a bit of work to make an rclone backend to that spec.

Fancy having a go?

**Comment by Rick7C2** [comment]
I would really like to see this feature! Unfortunately I don't know go very well so will be unable to help but I hope someone who can add this.

I would even be willing to start a bounty if enough people will add to the pool just to get this feature added to rclone.

**Comment by gary-kim** [comment]
It looks like the [referenced issue regarding minio taking a more library approach](https://github.com/minio/minio/issues/4812) seems to be closed saying that enough sub-systems have been moved to their own packages.

**Comment by pmjohann** [comment]
> Gofakes looks quite promising, eg the [backend definiton](https://godoc.org/github.com/johannesboyne/gofakes3#Backend).
> 
> It would be possible with a bit of work to make an rclone backend to that spec.
> 
> Fancy having a go?

+1 on this, as S3 being a (or THE) popular object storage protocol, I think this would benefit many.
Lots of applications offer S3 as a storage option but not others (e.g. FTP, SSH, WebDav etc. currently supported by rclone).

I also agree that there is no need to pull off 100% compatibility and go the minio route, as ACL & stuff like that are not required for most usecases.

**Comment by fnkr** [comment]
I'm surprised no one mentioned Minio Gateway yet, which does what "rclone serve s3" is supposed to do (except fewer backends supported): https://docs.min.io/docs/minio-gateway-for-s3.html

**Comment by gary-kim** [comment]
> I'm surprised no one mentioned Minio Gateway yet, which does what "rclone serve s3" is supposed to do (except fewer backends supported): https://docs.min.io/docs/minio-gateway-for-s3.html

It was mentioned

>     * https://github.com/minio/minio (100% compatible?)

The thing is, for example, I'm in a situation where I want to have an s3 compatible api way to access storage that is actually WebDAV. That is a thing that is pretty much impossible at the moment (without fuse that is, which isn't really reliable enough for me). Admittedly, it may end up being easier to add WebDAV support to Minio then S3 support to Rclone which would also mean that you can then chain Minio and Rclone to access any Rclone remote with `serve webdav` but in any case, adding `serve s3` support is one of the options, and in my opinion, the better one.

**Comment by ncw** [comment]
@gary-kim wrote
> but in any case, adding `serve s3` support is one of the options, and in my opinion, the better one.

It might be worth investing a bit of time into experimenting with an`rclone serve` with the gofakes library. The interface looked OK when I looked at it last so I don't think it would take too long to make a prototype. Fancy trying that @gary-kim ?

**Comment by gary-kim** [comment]
I'll try to find the time :)

**Comment by fionera** [comment]
A friend asked me if I want to implement this, so I hacked this together: https://github.com/fionera/rclone/tree/serve-minio

Let me know if this is ok for a PR. Because of the library the rclone binary goes up to 79MB when compiled :(

**Comment by unixfox** [comment]
@fionera great work!
Would you mind explaining how to test your code?
For instance what kind of environment variable should be set in order to get "serve minio" working?

**Comment by fionera** [comment]
I tested it a bit and noticed that you cannot upload files bigger than 200M or when you override the cutoff size not bigger than 5G since Multipart Uploads are not implemented. I will check how much work it is and maybe implement it. Also in the current version you need to add `--s3-no-check-bucket`, else rclone will try to create the bucket. This is not implemented (I fixed it locally by adding a check if it is the hardcoded one and not returning an error if so, tho I didnt push it yet).

**Comment by ncw** [comment]
> A friend asked me if I want to implement this, so I hacked this together: https://github.com/fionera/rclone/tree/serve-minio

:-)
 
> Let me know if this is ok for a PR. Because of the library the rclone binary goes up to 79MB when compiled :(

Ouch! I spent some time getting rid of the minio deps recently to reduce the binary size db56b1bfeceb7eaac6afe5e67bfff2d7d0e44b5d (I did the goftp refactoring too!).

It is a really nice feature though :-)

Do you think with a bit of work we could reduce the dependencies a bit? I note we import the top level "github.com/minio/minio/cmd" - I suspect that is the one that is bringing in all the dependencies. Maybe we could borrow a bit of code from there and not have to import it?

> I tested it a bit and noticed that you cannot upload files bigger than 200M or when you override the cutoff size not bigger than 5G since Multipart Uploads are not implemented. I will check how much work it is and maybe implement it.

OK

> Also in the current version you need to add `--s3-no-check-bucket`, else rclone will try to create the bucket. This is not implemented (I fixed it locally by adding a check if it is the hardcoded one and not returning an error if so, tho I didnt push it yet).

OK.

**Comment by stevefan1999-personal** [comment]
> I tested it a bit and noticed that you cannot upload files bigger than 200M or when you override the cutoff size not bigger than 5G since Multipart Uploads are not implemented. I will check how much work it is and maybe implement it. Also in the current version you need to add `--s3-no-check-bucket`, else rclone will try to create the bucket. This is not implemented (I fixed it locally by adding a check if it is the hardcoded one and not returning an error if so, tho I didnt push it yet).

would you like if i push a PR using https://golang.org/pkg/mime/multipart/, or maybe as an experiment

this is very useful especially you can leverage empty free storage as backup endpoint without having the quirks of FUSE and WebDAV

**Comment by unixfox** [comment]
> > I tested it a bit and noticed that you cannot upload files bigger than 200M or when you override the cutoff size not bigger than 5G since Multipart Uploads are not implemented. I will check how much work it is and maybe implement it. Also in the current version you need to add `--s3-no-check-bucket`, else rclone will try to create the bucket. This is not implemented (I fixed it locally by adding a check if it is the hardcoded one and not returning an error if so, tho I didnt push it yet).
> 
> would you like if i push a PR using [golang.org/pkg/mime/multipart](https://golang.org/pkg/mime/multipart/), or maybe as an experiment
> 
> this is very useful especially you can leverage empty free storage as backup endpoint without having the quirks of FUSE and WebDAV

Personally I would be very interested!

**Comment by Rick7C2** [comment]
> 
> 
> > I tested it a bit and noticed that you cannot upload files bigger than 200M or when you override the cutoff size not bigger than 5G since Multipart Uploads are not implemented. I will check how much work it is and maybe implement it. Also in the current version you need to add `--s3-no-check-bucket`, else rclone will try to create the bucket. This is not implemented (I fixed it locally by adding a check if it is the hardcoded one and not returning an error if so, tho I didnt push it yet).
> 
> would you like if i push a PR using https://golang.org/pkg/mime/multipart/, or maybe as an experiment
> 
> this is very useful especially you can leverage empty free storage as backup endpoint without having the quirks of FUSE and WebDAV

Curious if you had a chance to work on this?

It seems after multipart and the existing bucket check is implemented, then the only thing left to do is try and optimize the import so rclone binary is nice and lean correct?

@ncw I understand the need to keep the rclone binary size low and if the code imports can't be optimized could instructions on how to build with --serve s3 be added to docs for those who need this feature and are willing to have the extra file size?

Also is it safe to assume that rclone's caching abilities will "just work" with this new serve s3 feature or would that require more work?

**Comment by ivandeex** [comment]
Marking as important after 22 :+1: in total

**Comment by HeroCC** [comment]
@fionera I notice your rclone fork seem to no longer exist. Would you be willing to put it back up so we could take a peek at the work you've done so far?

**Comment by Mikubill** [comment]
I implemented a _basic_ s3 serve with gofakes3, currently it only supports limited functions (put, get, delete, list, etc.) but it works :) https://github.com/Mikubill/rclone/tree/serve-s3

**Comment by rkzed94** [comment]
@Mikubill, nice work!
is there any pre-compilled binary we could try instead compiling from source?
i dont know how to compile it by myself :/

**Comment by ncw** [comment]
@Mikubill great work :-) I hope you are going to send this as a PR at some point?

Hopefully you can get it working well enough so you can run rclone's s3 test suite against it and it passes - that would be the gold standardard!

**Comment by Rick7C2** [comment]
I assume @Mikubill will not be submitting a PR or running the test suite against it. 
Would anyone else be up to it?

I'd hate for this feature to be scrapped twice.

**Comment by remram44** [comment]
This is implemented, should the issue be closed? https://rclone.org/commands/rclone_serve_s3/

**Comment by ncw** [comment]
Yes it should be closed - thanks @remram44 

This feature got merged as `rclone serve s3` in v1.65.0

Any problems found, please make a new issue.
